### PR TITLE
added workaround scrollIntoView to center  to moveTo and click with a…

### PR DESCRIPTION
**Pull request by udarrr**
related to https://github.com/webdriverio/webdriverio/issues/11544

### CODE REVIEW

**Comment by erwinheitzman** [comment]
Please see my reply here: https://github.com/webdriverio/webdriverio/issues/11544

File: packages/webdriverio/src/commands/element/click.ts [hunk 1]
```diff
@@ -123,17 +123,30 @@ export async function click(
...
+        const clickNested = async () => {
+            await browser.action('pointer', {
+                parameters: { pointerType: 'mouse' }
             })
-            .down({ button })
-            .up({ button })
-            .perform(skipRelease)
+                .move({
+                    origin: this,
+                    x: xoffset,
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/click.ts:132]:
```suggestion
                    x: xoffset || 0,
```

File: packages/webdriverio/src/commands/element/moveTo.ts [hunk 2]
```diff
@@ -23,26 +21,29 @@ type MoveToOptions = {
...
     /**
      * W3C way of handle the mouse move actions
      */
     const browser = getBrowserObject(this)
-    await this.scrollIntoView({ block : 'nearest', inline: 'nearest', behavior: 'instant' })
-    if (!(await isIntoView())) {
-        await this.scrollIntoView({ block : 'center', inline: 'center', behavior: 'instant' })
+    const moveToNested = async ( xOffset?: number, yOffset?: number ) => {
+        return  await browser.action('pointer', { parameters: { pointerType: 'mouse' } })
+            .move({ origin: this, x: xOffset ? xOffset : 0, y: yOffset ? yOffset : 0 })
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:35]:
```suggestion
            .move({ origin: this, x: xOffset || 0, y: yOffset || 0 })
```

File: packages/webdriverio/src/commands/element/click.ts (hunk 1 above)
Comment (erwinheitzman) [reply packages/webdriverio/src/commands/element/click.ts:132]: Let's also rename xoffset to xOffset to keep it aligned with the moveTo command

File: packages/webdriverio/src/commands/element/click.ts [hunk 3]
```diff
@@ -123,17 +123,30 @@ export async function click(
...
+                    y: yoffset
```
[the remark this answers is not in the document]
Comment (erwinheitzman) [reply packages/webdriverio/src/commands/element/click.ts:133]: Let's also rename yoffset to yOffset to keep it aligned with the moveTo command

File: packages/webdriverio/src/commands/element/moveTo.ts [hunk 4]
```diff
@@ -23,26 +21,29 @@ type MoveToOptions = {
...
+            .perform()
+    }
+    try {
+        return await moveToNested(xOffset, yOffset)
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:39]:
```suggestion
        await moveToNested(xOffset, yOffset)
```

File: packages/webdriverio/src/commands/element/moveTo.ts [hunk 5]
```diff
@@ -23,26 +21,29 @@ type MoveToOptions = {
...
+    } catch {
+        /**
+        * Workaround, because sometimes browser.action().move() flaky and isn't able to scroll pointer to into view
+        * Moreover the action  with 'nearest' behavior by default where element is aligned at the bottom of its ancestor.
+        * and could be overlapped. Scroll to center should definitely work even if element was covered with sticky header/footer
+        */
+        await this.scrollIntoView({ block: 'center', inline: 'center', behavior: 'auto' })
+        return await moveToNested(xOffset, yOffset)
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:47]:
```suggestion
        return moveToNested(xOffset, yOffset)
```

File: packages/webdriverio/src/commands/element/click.ts (hunk 1 above)
Comment (erwinheitzman) [reply packages/webdriverio/src/commands/element/click.ts:132]: Actually, we can use the same implementation for setting the integers in the moveTo command. Again making the two commands more alike, in which case we do not need to cast it to 0 anymore.

File: packages/webdriverio/src/commands/element/click.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (erwinheitzman) [reply packages/webdriverio/src/commands/element/click.ts:133]: Actually, we can use the same implementation for setting the integers in the moveTo command. Again making the two commands more alike, in which case we do not need to cast it to 0 anymore.

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 2 above)
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:35]: Remove cast to integer and use click implementation

File: packages/webdriverio/src/commands/element/click.ts [hunk 6]
```diff
@@ -123,17 +123,30 @@ export async function click(
...
+        }
+        try {
+            await clickNested()
+        } catch {
+        /**
+        * Workaround, because sometimes browser.action().move() flaky and isn't able to scroll pointer to into view
+        * Moreover the action  with 'nearest' behavior by default where element is aligned at the bottom of its ancestor.
+        * and could be overlapped. Scroll to center should definitely work even if element was covered with sticky header/footer
+        */
+            await this.scrollIntoView({ block: 'center', inline: 'center', behavior: 'auto' })
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/click.ts:147]:
```suggestion
            await this.scrollIntoView({ block: 'center', inline: 'center' })
```

File: packages/webdriverio/src/commands/element/moveTo.ts [hunk 7]
```diff
@@ -23,26 +21,29 @@ type MoveToOptions = {
...
+    }
+    try {
+        return await moveToNested(xOffset, yOffset)
+    } catch {
+        /**
+        * Workaround, because sometimes browser.action().move() flaky and isn't able to scroll pointer to into view
+        * Moreover the action  with 'nearest' behavior by default where element is aligned at the bottom of its ancestor.
+        * and could be overlapped. Scroll to center should definitely work even if element was covered with sticky header/footer
+        */
+        await this.scrollIntoView({ block: 'center', inline: 'center', behavior: 'auto' })
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:46]:
```suggestion
        await this.scrollIntoView({ block: 'center', inline: 'center' })
```

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 4 above)
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/element/moveTo.ts:39]:
```suggestion
       await moveToNested(xOffset, yOffset)
```

**Review by erwinheitzman — CHANGES_REQUESTED** [review]

**Comment by erwinheitzman** [comment]
Thanks for the quick change @udarrr although I replied to the issue while you created the PR, I have also reviewed your work accordingly.

Great work 👍

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 2 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/moveTo.ts:35]: What do you mean by cat to integer ?

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 4 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/moveTo.ts]: done

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 7 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/moveTo.ts]: done

File: packages/webdriverio/src/commands/element/click.ts (hunk 6 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/click.ts]: done

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 2 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/moveTo.ts]: done

File: packages/webdriverio/src/commands/element/moveTo.ts (hunk 5 above)
Comment (udarrr) [reply packages/webdriverio/src/commands/element/moveTo.ts]: Did you mean 'return'  here ?

**Comment by udarrr** [comment]
> Thanks for the quick change @udarrr although I replied to the issue while you created the PR, I have also reviewed your work accordingly.
> 
> Great work 👍

That's great @erwinheitzman  thank you for supporting. I've done all changes related to your review.

**Review by erwinheitzman — CHANGES_REQUESTED** [review]

File: e2e/wdio/headless/test.e2e.ts [hunk 8]
```diff
@@ -186,6 +186,36 @@ describe('main suite 1', () => {
...
             expect(value.endsWith('center\n')).toBe(true)
         })
+
+        it('xOffset would cause a out of bounds', async () => {
+            try {
+                const elemRect = await browser.getElementRect(await browser.$('#parent').elementId)
+                await browser.$('#parent').moveTo({ xOffset: elemRect.width/2 + 1, yOffset: 0 })
+                throw new Error('xOffset and yOffset wouldn\'t cause a out of bounds')
+            } catch (err: any){
+                expect(err.message).toContain('xOffset would cause a out of bounds error as it goes outside of element')
```
[the remark this answers is not in the document]
Comment (udarrr) [reply e2e/wdio/headless/test.e2e.ts:196]: done

**Comment by erwinheitzman** [comment]
Loooking good but the tests could be better. Thank you for your contributions 😃

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

**Comment by christian-bromann** [comment]
> Loooking good but the tests could be better.

@erwinheitzman how can they be better? Maybe @udarrr can submit some test enhancements later on 🤔 

Thanks @udarrr for your contribution 🎉

**Comment by erwinheitzman** [comment]
@christian-bromann seems that by the time I added the comment, my feedback was already resolved. They worked at lightspeed 😁

**Pull request merged**